hack/cnbi-import.yaml

```diff
@@ -193,3 +194,173 @@ spec:
+        echo -n "    'jupyterhub' package is present"
+        if python -c "import jupyterhub" 2>/dev/null; then
+          ok
+        else
+          fail "error" "Missing 'jupyterhub' python package"
+        fi
+        echo -n "    'jupyterlab' package is present"
+        if python -c "import jupyterlab" 2>/dev/null; then
+          ok
+        else
+          fail "error" "Missing 'jupyterlab' python package"
+        fi
+
+        # HOME env var is present and points to writeable directory
+        echo -n "    \$HOME is writeable"
+        if [ -n "$HOME" ] && [ -w "$HOME" ]; then
+          ok
+        else
+          fail "error" "\$HOME has to be set and point to writeable directory"
+        fi
+
+        # start-singleuser.sh must be present and resolve to "jupyter labhub $@"
+
+        function jupyter {
+          echo "$@" > /tmp/entrypoint.txt
+        }
+        export -f jupyter
+        export JUPYTERHUB_API_TOKEN="test"
+
+        echo -n "    start-singleuser.sh is present "
+        if ! command -v start-singleuser.sh >/dev/null; then
+          fail "error" "start-singleuser.sh script must be present in $PATH and executable"
+        else
+          ok
+        fi
+
+        echo -n "    start-singleuser.sh must execute 'jupyter' "
+        if ! timeout 60 start-singleuser.sh --ip=0.0.0.0 --port=8080 >/dev/null 2>/dev/null; then
+            fail "error" "start-singleuser.sh script took too long to call 'jupyter labhub'"
+        else
+          if [ ! -f /tmp/entrypoint.txt ]; then
+            fail "error" "start-singleuser.sh script did not call 'jupyter'"
+          else
+            ok
+          fi
+        fi
+
+        echo -n "    start-singleuser.sh must start 'labhub' environment "
+        if [ ! -f /tmp/entrypoint.txt ]; then
+          # This error is already raised by a previous check
+          skip
+        else
+          if grep -v "^labhub" /tmp/entrypoint.txt >/dev/null 2>/dev/null; then
+            fail "error" "start-singleuser.sh script did not call 'jupyter labhub'"
+          else
+            ok
+          fi
+        fi
+
+        echo -n "    start-singleuser.sh must accept runtime args or pass a config file "
+        if [ ! -f /tmp/entrypoint.txt ]; then
+          # This error is already raised by a previous check
+          skip
+        else
+          if grep "\$@" /tmp/entrypoint.txt >/dev/null 2>/dev/null; then
+            ok
+          else
+            if grep -- "--port=8080" /tmp/entrypoint.txt >/dev/null 2>/dev/null; then
+              ok
+            else
+              config=$(grep -oP "(?<=--config=)\S*" /tmp/entrypoint.txt 2>/dev/null)
+              if [ -n "$config" ]; then
+                fail "error" "start-singleuser.sh script has to either use runtime arguments via '\$@' or specify '--port=8080' or it needs to specify a custom config file."
+              else
+                ok
+              fi
+            fi
+          fi
+        fi
```

Comment:
Isn't implementing all that flow control in bash inline in the tekton spec be hard to maintain in the long run ?

Maybe we could either use Tekton pipeline strucure as a control flow, or implement this in an image  and use the param.url as data (not sure is there is a tool to mount a image as a pod volume though) ?

Reply:
I think @tumido created this part (I didnt have a look at `git blame`) and I think you are right, it should be optimized:

- controll flow is a matter of tekton
- scripts/application/... should be callables in container images
- reducing the mixture of bash, yaml, python should be a goal, as it will increase maintainability

wdyt?

Reply:
If you are looking for `git praise`, the source of the pipelines is in the [helm-charts repo](https://github.com/thoth-station/helm-charts/) (see the [import](https://github.com/thoth-station/helm-charts/blob/08e78c0bd12cfe85cad65d5d1b6979ddb45ee1fb/charts/meteor-pipelines/templates/byon-import-jupyterhub-image.yaml) and [build](https://github.com/thoth-station/helm-charts/blob/08e78c0bd12cfe85cad65d5d1b6979ddb45ee1fb/charts/meteor-pipelines/templates/meteor_jupyterhub.yaml) pipelines there, and associated task manifests too)

There are indeed things to improve in the pipelines. That flow is probably one of them - and especially because it might need reviewing to work with the Kubeflow notebook controller. Another one is the selection of the base image. Also: how/if we handle metrics, how/where to incorporate Thoth advise...

This PR focuses in the controller part, and I would personally consider it ready when the controller does create the "right" pipeline given a `GitRepository` build type. All the pipeline improvements could/should be handled in other issues/PRs.

Makes sense?

Reply:
It does. I hadn't realise this was taken from another repo at first ^ :+1: